### ISSUE TITLE
Could not replicate results as given in paper

### ISSUE THREAD

**Issue by Rellik-7** [shirley-wu/text_to_table#7]
Hi! Great work.

I followed the instructions as mentioned in the readme on wikitabletext data, and wasn't able to replicate the results. I trained a HAD model, and ran inference using the test_constraint script but got inferior results in the BertScore metric. (Row header: 89.53 Non-header: 74.33) which are nearly 6 points lower than the scores in the paper.

Also, during training, my best models seem to get about 2.6 loss but an earlier issue mentioned that a 2.26 loss is expected.

Log: checkpoints//checkpoint26.best_2.6280.pt checkpoints//checkpoint24.best_2.6270.pt checkpoints//checkpoint22.best_2.6270.pt

Would you have any idea where I am going wrong?

**Comment by shirley-wu** [comment]
Hi, thanks for your interests in this work!

~2.6 loss is ok. I got ~2.6 loss in my experiments too. Sorry that my comments on the other issue is confusing.

For BERTScore, I got the same problem as you and I'm not yet sure what's wrong. I rerun the evaluation scripts on my previous outputs and got different BERTScore numbers for all experiments (for wikitabletext, it's 74~76, as in your experiment). For the other two metrics I got the exact same numbers as in the paper. Since I re-configured my python environment (my previous environment was gone) I'm suspecting there's some issue with bert_score package. Will get back when I figure it out.

What's the number for other two metrics in your experiments? If they're behaving normally you can just use these two metrics. Sorry for this.

**Comment by shirley-wu** [comment]
Hi, I have found the reason for this bug. **The short answer is, please use `transformers<4.17.0`.** For example, I use `transformers==3.1.0` and have obtained the exact same numbers as in the paper.

The detailed explanations are as follows.

In bert_score package, when tokenizing the sentence, a prefix space is added to the sentence for GPT2Tokenizer ([code](https://github.com/Tiiiger/bert_score/blob/v0.3.10/bert_score/utils.py#L184)). In `transformers<4.17.0`, RobertaTokenizer is a subclass of GPT2Tokenizer ([code](https://github.com/huggingface/transformers/blob/v4.16.2/src/transformers/models/roberta/tokenization_roberta.py#L60)), so the prefix space is added. Starting from `transformers==4.17.0`, RobertaTokenizer is no longer a subclass of GPT2Tokenizer ([code](https://github.com/huggingface/transformers/blob/v4.17.0/src/transformers/models/roberta/tokenization_roberta.py#L103)), so the prefix space is not added. I am not sure why adding a prefix space will cause such a dramatic change in the final scores. But according to the comments in bert_score source code, the prefix space should be added for RoBERTa and GPT2, so I believe using `transformers<4.17.0` will yield the correct scores.

I will update the readme file and `requirements.txt`. Please let me know if this cannot resolve your problem.

**Comment by tranluongtiensi** [comment]
excuse me, You can tell me how to train model to have a *pt file in checkpoints ? I also run bash command like readme, but it show me error :
2023-06-11 23:00:05 | INFO | fairseq.trainer | begin training epoch 1
ls: cannot access 'checkpoints//*pt': No such file or directory
ls: cannot access 'checkpoints//checkpoint*best_*pt': No such file or directory

usage: average_checkpoints.py [-h] --inputs INPUTS [INPUTS ...] --output FILE [--num-epoch-checkpoints NUM_EPOCH_CHECKPOINTS | --num-update-checkpoints NUM_UPDATE_CHECKPOINTS]
                              [--checkpoint-upper-bound CHECKPOINT_UPPER_BOUND]
average_checkpoints.py: error: argument --inputs: expected at least one argument

the error happen when I run command `bash scripts/rotowire/train_had.sh data/rotowire/ bart.base/`

**Comment by shirley-wu** [comment]
@tranluongtiensi hi, this is because the training script `custom_train.py` failed. Could you post the training log here so I can check what's wrong? It's supposed to be at `checkpoints/log`

**Comment by tranluongtiensi** [comment]
@shirley-wu hi, this is my training log
2023-06-12 23:13:33 | INFO | fairseq_cli.train | Namespace(activation_fn='gelu', adam_betas='(0.9, 0.999)', adam_eps=1e-08, adaptive_softmax_cutoff=None, adaptive_softmax_dropout=0, all_gather_list_size=16384, arch='bart_base', attention_dropout=0.1, batch_size=None, batch_size_valid=None, best_checkpoint_metric='loss', bf16=False, bpe=None, broadcast_buffers=False, bucket_cap_mb=25, checkpoint_shard_count=1, checkpoint_suffix='', clip_norm=0.1, cpu=False, criterion='label_smoothed_cross_entropy', cross_self_attention=False, curriculum=0, data='data/rotowire//bins', data_buffer_size=10, dataset_impl=None, ddp_backend='c10d', decoder_attention_heads=12, decoder_embed_dim=768, decoder_embed_path=None, decoder_ffn_embed_dim=3072, decoder_input_dim=768, decoder_layerdrop=0, decoder_layers=6, decoder_layers_to_keep=None, decoder_learned_pos=True, decoder_normalize_before=False, decoder_output_dim=768, device_id=0, disable_validation=False, distributed_backend='nccl', distributed_init_method=None, distributed_no_spawn=False, distributed_num_procs=0, distributed_port=-1, distributed_rank=0, distributed_world_size=1, distributed_wrapper='DDP', dropout=0.1, empty_cache_freq=0, encoder_attention_heads=12, encoder_embed_dim=768, encoder_embed_path=None, encoder_ffn_embed_dim=3072, encoder_layerdrop=0, encoder_layers=6, encoder_layers_to_keep=None, encoder_learned_pos=True, encoder_normalize_before=False, eval_bleu=False, eval_bleu_args=None, eval_bleu_detok='space', eval_bleu_detok_args=None, eval_bleu_print_samples=False, eval_bleu_remove_bpe=None, eval_tokenized_bleu=False, fast_stat_sync=False, find_unused_parameters=True, finetune_from_model=None, fix_batches_to_gpus=False, fixed_validation_seed=None, fp16=False, fp16_init_scale=128, fp16_no_flatten_grads=False, fp16_scale_tolerance=0.0, fp16_scale_window=None, gen_subset='test', ignore_prefix_size=0, keep_best_checkpoints=3, keep_interval_updates=-1, keep_last_epochs=-1, label_smoothing=0.1, layernorm_embedding=True, left_pad_source='True', left_pad_target='False', load_alignments=False, localsgd_frequency=3, log_format=None, log_interval=100, lr=[3e-05], lr_scheduler='inverse_sqrt', max_epoch=0, max_source_positions=1024, max_target_positions=1024, max_tokens=4096, max_tokens_valid=4096, max_update=8000, maximize_best_checkpoint_metric=False, memory_efficient_bf16=False, memory_efficient_fp16=False, min_loss_scale=0.0001, min_lr=-1.0, model_parallel_size=1, no_cross_attention=False, no_epoch_checkpoints=False, no_last_checkpoints=False, no_progress_bar=False, no_save=False, no_save_optimizer_state=False, no_scale_embedding=True, no_seed_provided=False, no_token_positional_embeddings=False, nprocs_per_node=1, num_batch_buckets=0, num_shards=1, num_workers=8, optimizer='adam', optimizer_overrides='{}', patience=-1, pipeline_balance=None, pipeline_checkpoint='never', pipeline_chunks=0, pipeline_decoder_balance=None, pipeline_decoder_devices=None, pipeline_devices=None, pipeline_encoder_balance=None, pipeline_encoder_devices=None, pipeline_model_parallel=False, pooler_activation_fn='tanh', pooler_dropout=0.0, profile=False, quant_noise_pq=0, quant_noise_pq_block_size=8, quant_noise_scalar=0, quantization_config_path=None, relu_dropout=0.0, report_accuracy=False, required_batch_size_multiple=1, required_seq_len_multiple=1, reset_dataloader=True, reset_lr_scheduler=False, reset_meters=True, reset_optimizer=True, restore_file='bart.base//model.pt', save_dir='checkpoints', save_interval=10, save_interval_updates=0, scoring='bleu', seed=10, sentence_avg=False, shard_id=0, share_all_embeddings=True, share_decoder_input_output_embed=True, skip_invalid_size_inputs_valid_test=True, slowmo_algorithm='LocalSGD', slowmo_momentum=None, source_lang='text', stop_time_hours=0, target_lang='data', task='translation', tensorboard_logdir=None, threshold_loss_scale=None, tokenizer=None, tpu=False, train_subset='train', truncate_source=True, update_freq=[1], upsample_primary=1, use_bmuf=False, use_old_adam=False, user_dir=None, valid_subset='valid', validate_after_updates=0, validate_interval=10, validate_interval_updates=0, warmup_init_lr=1e-07, warmup_updates=400, weight_decay=0.01, zero_sharding='none')
2023-06-12 23:13:33 | INFO | fairseq.tasks.translation | [text] dictionary: 51200 types
2023-06-12 23:13:33 | INFO | fairseq.tasks.translation | [data] dictionary: 51200 types
2023-06-12 23:13:33 | INFO | fairseq.data.data_utils | loaded 727 examples from: data/rotowire//bins/valid.text-data.text
2023-06-12 23:13:33 | INFO | fairseq.data.data_utils | loaded 727 examples from: data/rotowire//bins/valid.text-data.data
2023-06-12 23:13:33 | INFO | fairseq.tasks.translation | data/rotowire//bins valid text-data 727 examples
2023-06-12 23:13:36 | INFO | fairseq_cli.train | BARTModel(
  (encoder): TransformerEncoder(
    (dropout_module): FairseqDropout()
    (embed_tokens): Embedding(51200, 768, padding_idx=1)
    (embed_positions): LearnedPositionalEmbedding(1026, 768, padding_idx=1)
    (layernorm_embedding): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
    (layers): ModuleList(
      (0-5): 6 x TransformerEncoderLayer(
        (self_attn): MultiheadAttention(
          (dropout_module): FairseqDropout()
          (k_proj): Linear(in_features=768, out_features=768, bias=True)
          (v_proj): Linear(in_features=768, out_features=768, bias=True)
          (q_proj): Linear(in_features=768, out_features=768, bias=True)
          (out_proj): Linear(in_features=768, out_features=768, bias=True)
        )
        (self_attn_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (dropout_module): FairseqDropout()
        (activation_dropout_module): FairseqDropout()
        (fc1): Linear(in_features=768, out_features=3072, bias=True)
        (fc2): Linear(in_features=3072, out_features=768, bias=True)
        (final_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      )
    )
  )
  (decoder): TransformerDecoder(
    (dropout_module): FairseqDropout()
    (embed_tokens): Embedding(51200, 768, padding_idx=1)
    (embed_positions): LearnedPositionalEmbedding(1026, 768, padding_idx=1)
    (layernorm_embedding): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
    (layers): ModuleList(
      (0-5): 6 x TransformerDecoderLayer(
        (dropout_module): FairseqDropout()
        (self_attn): MultiheadAttention(
          (dropout_module): FairseqDropout()
          (k_proj): Linear(in_features=768, out_features=768, bias=True)
          (v_proj): Linear(in_features=768, out_features=768, bias=True)
          (q_proj): Linear(in_features=768, out_features=768, bias=True)
          (out_proj): Linear(in_features=768, out_features=768, bias=True)
        )
        (activation_dropout_module): FairseqDropout()
        (self_attn_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (encoder_attn): MultiheadAttention(
          (dropout_module): FairseqDropout()
          (k_proj): Linear(in_features=768, out_features=768, bias=True)
          (v_proj): Linear(in_features=768, out_features=768, bias=True)
          (q_proj): Linear(in_features=768, out_features=768, bias=True)
          (out_proj): Linear(in_features=768, out_features=768, bias=True)
        )
        (encoder_attn_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (fc1): Linear(in_features=768, out_features=3072, bias=True)
        (fc2): Linear(in_features=3072, out_features=768, bias=True)
        (final_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      )
    )
    (output_projection): Linear(in_features=768, out_features=51200, bias=False)
  )
  (classification_heads): ModuleDict()
)
2023-06-12 23:13:36 | INFO | fairseq_cli.train | task: translation (TranslationTask)
2023-06-12 23:13:36 | INFO | fairseq_cli.train | model: bart_base (BARTModel)
2023-06-12 23:13:36 | INFO | fairseq_cli.train | criterion: label_smoothed_cross_entropy (LabelSmoothedCrossEntropyCriterion)
2023-06-12 23:13:36 | INFO | fairseq_cli.train | num. model params: 140138496 (num. trained: 140138496)
2023-06-12 23:13:36 | INFO | fairseq.trainer | detected shared parameter: encoder.embed_tokens.weight <- decoder.embed_tokens.weight
2023-06-12 23:13:36 | INFO | fairseq.trainer | detected shared parameter: encoder.embed_tokens.weight <- decoder.output_projection.weight
2023-06-12 23:13:36 | INFO | fairseq_cli.train | training on 1 devices (GPUs/TPUs)
2023-06-12 23:13:36 | INFO | fairseq_cli.train | max tokens per GPU = 4096 and max sentences per GPU = None
2023-06-12 23:13:38 | INFO | fairseq.trainer | loaded checkpoint bart.base//model.pt (epoch 14 @ 0 updates)
2023-06-12 23:13:38 | INFO | fairseq.trainer | loading train data for epoch 1
2023-06-12 23:13:38 | INFO | fairseq.data.data_utils | loaded 3398 examples from: data/rotowire//bins/train.text-data.text
2023-06-12 23:13:38 | INFO | fairseq.data.data_utils | loaded 3398 examples from: data/rotowire//bins/train.text-data.data
2023-06-12 23:13:38 | INFO | fairseq.tasks.translation | data/rotowire//bins train text-data 3398 examples
2023-06-12 23:13:39 | INFO | fairseq.trainer | begin training epoch 1

**Comment by shirley-wu** [comment]
@tranluongtiensi I don't see any error from this log. Is it possible that your training process got killed, such as it exceeds memory limit?

**Comment by tranluongtiensi** [comment]
@shirley-wu I running custom_train.py with this command:
python custom_train.py --num-workers 8 data/rotowire/bins --seed 10 --keep-best-checkpoints 3 --save-interval 10 --validate-interval 10 --restore-file bart.base/model.pt --max-tokens 4096 --task translation --source-lang text --target-lang data --truncate-source --layernorm-embedding --share-all-embeddings --share-decoder-input-output-embed --reset-optimizer --reset-dataloader --reset-meters --required-batch-size-multiple 1 --arch bart_base --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.01 --optimizer adam --adam-betas "(0.9, 0.999)" --adam-eps 1e-08 --clip-norm 0.1 --lr-scheduler inverse_sqrt --lr 3e-05 --max-update 8000 --warmup-updates 400 --warmup-init-lr '1e-07' --update-freq 1 --skip-invalid-size-inputs-valid-test --find-unused-parameters. 

Is it correct?

**Comment by shirley-wu** [comment]
@tranluongtiensi I think it's correct. It trains the baseline model rather than the HAD model. How does it go?

**Comment by tranluongtiensi** [comment]
@shirley-wu  it was killed, maybe problem like you said exceed memory limit. Maybe I should try in other computer with more memory. about trains HAD model, you can tell me  what command I should type on terminal ? I don't see you mention in the readme file

**Comment by shirley-wu** [comment]
@tranluongtiensi for running HAD, you should change `--arch bart_base` into `--arch bart_ours_base` and add `--user-dir src/ --return-relative-column-strs row_head col_head`

Also, note that if you want to run multiple training experiments, you would want to change the checkpointing directory `--save-dir` so they don't override each other
